itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/MetastoreHousekeepingLeaderTestBase.java

```diff
@@ -0,0 +1,192 @@
+    long numTasks = addRemoteOnlyTasksConfigs();
+    numTasks = numTasks + addAlwaysTasksConfigs();
+    MetastoreConf.setLongVar(conf, ConfVars.THREAD_POOL_SIZE, numTasks);
+  }
+  private void addStatsUpdaterThreadConfigs() {
+    MetastoreConf.setLongVar(conf, ConfVars.STATS_AUTO_UPDATE_WORKER_COUNT, 1);
+    MetastoreConf.setVar(conf, ConfVars.STATS_AUTO_UPDATE, "all");
+    threadClasses.put(StatsUpdaterThread.class, false);
+    threadNames.put(StatsUpdaterThread.WORKER_NAME_PREFIX, false);
+  }
+
+  private void addReplChangeManagerConfigs() throws Exception {
+    miniDFS = new MiniDFSCluster.Builder(new Configuration()).numDataNodes(1).format(true).build();
+    MetastoreConf.setBoolVar(conf, ConfVars.REPLCMENABLED, true);
+    String cmroot = "hdfs://" + miniDFS.getNameNode().getHostAndPort() + "/cmroot";
+    MetastoreConf.setVar(conf, ConfVars.REPLCMDIR, cmroot);
+    threadNames.put(ReplChangeManager.CM_THREAD_NAME_PREFIX,  false);
+  }
+
+  private void addCompactorConfigs() {
+    MetastoreConf.setBoolVar(conf, ConfVars.COMPACTOR_INITIATOR_ON, true);
+    MetastoreConf.setVar(conf, ConfVars.HIVE_METASTORE_RUNWORKER_IN, "metastore");
+    MetastoreConf.setLongVar(conf, ConfVars.COMPACTOR_WORKER_THREADS, 1);
+    threadClasses.put(Initiator.class, false);
+    threadClasses.put(Worker.class, false);
+    threadClasses.put(Cleaner.class, false);
+  }
+
+  private long addRemoteOnlyTasksConfigs() {
+    String remoteTaskClassPaths =
+            MetastoreTaskThreadTestImpl1.class.getCanonicalName() + "," +
+                    MetastoreTaskThreadTestImpl2.class.getCanonicalName();
+
+    MetastoreConf.setVar(conf, ConfVars.TASK_THREADS_REMOTE_ONLY, remoteTaskClassPaths);
+
+    threadNames.put(MetastoreTaskThreadTestImpl1.TASK_NAME, false);
+    threadNames.put(MetastoreTaskThreadTestImpl2.TASK_NAME, false);
+
+    return 2;
+  }
+
+  private long addAlwaysTasksConfigs() throws Exception {
+    String alwaysTaskClassPaths = MetastoreTaskThreadTestImpl3.class.getCanonicalName();
+    MetastoreConf.setVar(conf, ConfVars.TASK_THREADS_ALWAYS, alwaysTaskClassPaths);
+    threadNames.put(MetastoreTaskThreadTestImpl3.TASK_NAME, false);
+    return 1;
+  }
+
+  private static String getAllThreadsAsString() {
+    Map<Thread, StackTraceElement[]> threadStacks = Thread.getAllStackTraces();
+    StringBuilder sb = new StringBuilder();
+    for (Map.Entry<Thread, StackTraceElement[]> entry : threadStacks.entrySet()) {
+      Thread t = entry.getKey();
+      sb.append(System.lineSeparator());
+      sb.append("Name: ").append(t.getName()).append(" State: ").append(t.getState())
+              .append(" Class name: ").append(t.getClass().getCanonicalName());
+    }
+    return sb.toString();
+  }
+
+  void searchHousekeepingThreads() throws Exception {
+    // Client has been created so the metastore has started serving. Sleep for few seconds for
+    // the housekeeping threads to start.
+    Thread.sleep(SLEEP_INTERVAL_FOR_THREADS_TO_START);
```

Comment:
sleep may cause flaky test if the system is slow.

Reply:
That's why I am adding a rather long delay like 10s. Usually the threads are started within a second after the metastore is started.

I don't want to add a synchronization just for testing to check whether threads are started.

We could add retry mechanism similar to HiveMetaStoreClient() which creates a connection to metastore. But we are already calling HiveMetaStoreClient() to ensure that metastore has started to serve (the condition after which threads are started). So, I don't see much point in retry as well. A sufficiently long delay should be enough.

Do you have any specific solution in mind?